src/runtime_src/core/common/usage_metrics.cpp

```diff
@@ -0,0 +1,515 @@
+                        {
+                          return met.handle == finder;
+                        }
+                        );
+  return it == metrics_vec.end()
+    ? nullptr
+    : &(*it);
+}
+
+struct bo_metrics
+{
+  uint32_t total_count = 0;
+  size_t   total_size_in_bytes = 0;
+  size_t   peak_size_in_bytes = 0;
+  size_t   bytes_synced_to_device = 0;
+  size_t   bytes_synced_from_device = 0;
+};
+
+using tp = std::chrono::time_point<std::chrono::high_resolution_clock>;
+struct kernel_metrics
+{
+  struct timestamp
+  {
+    tp start_time;
+    bool is_valid = false;
+  };
+
+  std::string handle; // kernel name is used as handle for identifying kernel
+  std::vector<uint32_t> cu_index_vec;
+  uint32_t total_runs = 0;
+  std::chrono::microseconds total_time = {};
+  std::unordered_map<const xrt::run_impl*, timestamp> exec_times; // run handle ptr is used for indexing
+  size_t num_args;
+
+  void
+  log_kernel_exec_time(const xrt::run_impl* run_hdl, const tp& tp_now, ert_cmd_state state)
+  {
+    // state ERT_CMD_STATE_NEW indicates kernel start is called
+    if (state == ERT_CMD_STATE_NEW) {
+      // record start everytime because previous run may be finished, timeout, aborted or stopped
+      exec_times[run_hdl].start_time = tp_now;
+      exec_times[run_hdl].is_valid = true;
+    }
+    else {
+      // make start time invalid so we can record for next run, add duration to total time and increment total runs
+      if (exec_times[run_hdl].is_valid && state == ERT_CMD_STATE_COMPLETED) {
+        // valid run increament run
+        total_runs++;
+        total_time += std::chrono::duration_cast<std::chrono::microseconds>(tp_now - exec_times[run_hdl].start_time);
+
+        // invalidate start time, run may be finished, aborted or timed out
+        exec_times[run_hdl].is_valid = false;
+      }
+    }
+  }
+};
+
+struct hw_ctx_metrics
+{
+  const xrt_core::hwctx_handle* handle;  // using hw_ctx handle ptr as unique identifier for logging
+  xrt::uuid xclbin_uuid;
+  struct bo_metrics bos_met;
+  std::vector<struct kernel_metrics> kernel_metrics_vec;
+
+  void
+  log_kernel(const std::string& name, size_t args)
+  {
+    struct kernel_metrics k;
+    k.handle = name;
+    k.num_args = args;
+    kernel_metrics_vec.emplace_back(k);
+  }
+};
+
+struct device_metrics
+{
+  std::string bdf = "";
+  struct bo_metrics global_bos_met;
+  uint32_t bo_active_count = 0;
+  uint32_t bo_peak_count = 0;
+  std::vector<struct hw_ctx_metrics> hw_ctx_vec;
+
+  void
+  log_hw_ctx(const xrt_core::hwctx_handle* handle, const xrt::uuid& uuid)
+  {
+    hw_ctx_vec.emplace_back(hw_ctx_metrics{handle, uuid, {}, {}});
+  }
+};
+
+using metrics_map = std::map<std::thread::id, std::map<device_id, device_metrics>>;
+using device_metrics_map = std::map<device_id, device_metrics>;
+// Global map to store usage metrics of all threads
+std::shared_ptr<metrics_map> usage_metrics_map = std::make_shared<metrics_map>();
```

Comment:
Definitely external linkage of variables in a named namespace.  @keryell this is why I always stick with static even in the unnamed namespace.  Here we for sure need static to avoid external linkage.
```suggestion
static std::shared_ptr<metrics_map> usage_metrics_map = std::make_shared<metrics_map>();
```

Reply:
DRY
```c++
static auto usage_metrics_map = std::make_shared<metrics_map>();
```
Could be an anonymous namespace instead. :-)

Reply:
Of course `auto`.    Yes, it could be in anonymous but to what advantage?

Reply:
Some people think this is good for some reasons https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sf22-use-an-unnamed-anonymous-namespace-for-all-internalnon-exported-entities

Reply:
I have not followed the discussion behind this in the ISO C++ committee but I guess `static` is a very overloaded C keyword with a lot of different things.
In C++, anonymous namespace is here to make very clear the meaning of the programmer instead of writing `static` where a reader has to think about what is *this* `static` means here.

Reply:
I agree with that.  I always use unnamed namespace for implementation details and only the named one for exported (per header) content.  I kind of missed that in this PR.  @rbramand-xilinx I think we should move everything in usage_metrics.cpp into the unamed namespace with the exception of the `get_usage_metrics_logger()` of course which the only external API.

Reply:
> In C++, anonymous namespace is here to make very clear the meaning of the programmer instead of writing `static` where a reader has to think about what is _this_ `static` means here.

Completely disagree.   You may have to search long and hard to realize that the definition is within an unnamed namespace and indeed has desired internal linkage.   Personally I will continue to use static even within unnamed, unless you tell me it has side effects beyond not needed.

Reply:
I understand you point here.
At the end this is just a coding guideline among others and why not using yours in your project. :-)